### PR TITLE
[IR] Remove non-canonical matchings

### PR DESCRIPTION
**Pull request by AZero13**

### CODE REVIEW

**Comment by nikic** [comment]
This is not NFC, and there are test failures.

**Comment by AtariDreams** [comment]
@nikic Is this good now?

**Review by RKSimon — COMMENTED** [review]
What are you trying to accomplish with this patch? The SLSR change looks sensible, we use the m_c_Add/m_c_Or matchers to avoid duplication in commutative ops. But what are you trying to avoid with the other changes? Are you certain that we won't see constants on the LHS on the ADD/AND ops? Canonicalization isn't always done in time - especially in general code like getKnownBitsFromAndXorOr.

**Review by nikic — APPROVED** [review]
LGTM from my side. Non-canonical patterns should only be handled if there is a specific need, in which case there should at minimum be a test case that justifies it, which is not present here.

File: llvm/lib/Transforms/Scalar/StraightLineStrengthReduce.cpp [hunk 1]
```diff
@@ -425,14 +425,12 @@ void StraightLineStrengthReduce::allocateCandidatesAndFindBasisForAdd(
...
-  return (match(A, m_Add(m_Value(B), m_ConstantInt(C))) ||
-          match(A, m_Add(m_ConstantInt(C), m_Value(B))));
+  return match(A, m_c_Add(m_Value(B), m_ConstantInt(C)));
 }
 
 // Returns true if A matches B | C where C is constant.
 static bool matchesOr(Value *A, Value *&B, ConstantInt *&C) {
-  return (match(A, m_Or(m_Value(B), m_ConstantInt(C))) ||
-          match(A, m_Or(m_ConstantInt(C), m_Value(B))));
+  return match(A, m_c_Or(m_Value(B), m_ConstantInt(C)));
```
Comment (goldsteinn) [inline llvm/lib/Transforms/Scalar/StraightLineStrengthReduce.cpp:433]: This seems to be adding support for non-canonical version...

Comment (goldsteinn) [inline llvm/lib/Transforms/Scalar/StraightLineStrengthReduce.cpp:433]: Do we need commutative version here/above?

Comment (AtariDreams) [reply llvm/lib/Transforms/Scalar/StraightLineStrengthReduce.cpp:433]: Yeah. Otherwise tests break.

**Comment by goldsteinn** [comment]
LGTM

**Pull request merged**